h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/BroadcastJoinForTargetEncoder.java

```diff
@@ -0,0 +1,174 @@
+            int valueFromRIghtArr = rightArr[rowIdx][colIdx];
+            leftArr[rowIdx][colIdx] = Math.max(valueFromLeftArr, valueFromRIghtArr);
+          }
+        }
+      }
+    }
+    
+    int[][] getEncodingDataArray() {
+      return _encodingDataPerNode;
+    }
+  }
+
+  /**
+   * 
+   * @param leftFrame frame that we want to keep rows order of
+   * @param leftCatColumnsIdxs indices of the categorical columns from `leftFrame` we want to use to calculate encodings
+   * @param leftFoldColumnIdx index of the fold column from `leftFrame` or `-1` if we don't use folds
+   * @param broadcastedFrame supposedly small frame that we will broadcast to all nodes and use it as lookup table for joining
+   * @param rightCatColumnsIdxs indices of the categorical columns from `broadcastedFrame` we want to use to calculate encodings 
+   * @param rightFoldColumnIdx index of the fold column from `broadcastedFrame` or `-1` if we don't use folds
+   * @return `leftFrame` with joined numerators and denominators
+   */
+  static Frame join(Frame leftFrame, int[] leftCatColumnsIdxs, int leftFoldColumnIdx, Frame broadcastedFrame, int[] rightCatColumnsIdxs, int rightFoldColumnIdx, int maxFoldValue) {
+    int numeratorIdx = broadcastedFrame.find(TargetEncoder.NUMERATOR_COL_NAME);
+    int denominatorIdx = broadcastedFrame.find(TargetEncoder.DENOMINATOR_COL_NAME);
+
+    int broadcastedFrameCatCardinality = broadcastedFrame.vec(rightCatColumnsIdxs[0]).cardinality();
+
+    if(rightFoldColumnIdx != -1 && broadcastedFrame.vec(rightFoldColumnIdx).max() > Integer.MAX_VALUE) 
+      throw new IllegalArgumentException("Fold value should be a non-negative integer (i.e. should belong to [0, Integer.MAX_VALUE] range)");
+      
+    int[][] levelMappings = {CategoricalWrappedVec.computeMap( leftFrame.vec(leftCatColumnsIdxs[0]).domain(), broadcastedFrame.vec(0).domain())};
+
+    int[][] encodingDataMap = new FrameWithEncodingDataToArray(rightCatColumnsIdxs[0], rightFoldColumnIdx, numeratorIdx, denominatorIdx, broadcastedFrameCatCardinality, maxFoldValue)
+            .doAll(broadcastedFrame)
+            .getEncodingDataArray();
+    new BroadcastJoiner(leftCatColumnsIdxs, leftFoldColumnIdx, encodingDataMap, levelMappings, broadcastedFrameCatCardinality)
+            .doAll(leftFrame);
+    return leftFrame;
+  }
+
+  static class BroadcastJoiner extends MRTask<BroadcastJoiner> {
+    int _categoricalColumnIdx, _foldColumnIdx, _cardinalityOfCatCol;
+    int[][] _encodingDataArray;
+    int[][] _levelMappings;
+
+    BroadcastJoiner(int[] categoricalColumnsIdxs, int foldColumnIdx, int[][] encodingDataMap, int[][] levelMappings, int cardinalityOfCatCol) {
+      assert categoricalColumnsIdxs.length == 1 : "Only single column target encoding(i.e. one categorical column is used to produce its encodings) is supported for now";
+
+      _categoricalColumnIdx = categoricalColumnsIdxs[0];
+      _foldColumnIdx = foldColumnIdx;
+      _encodingDataArray = encodingDataMap;
```

Comment:
name change confused me when reading code below

Reply:
@seb-h2o Yes, it is confusing. Will fix.